### ISSUE TITLE
Widevine version issue

### ISSUE THREAD

**Issue by cache-sk** [emilsvennesson/script.module.inputstreamhelper#59]
Some encrypted video sources stopped working today.
Problem is, helper uses old widevine version.
Yes, https://dl.google.com/widevine-cdm/current.txt says that current version is 1.4.9.1088 but that's not true.
https://dl.google.com/widevine-cdm/versions.txt shows bunch of versions, 4.10.1440.19 is latest. For example, official Chrome on windows x64 uses 4.10.1440.18.

For me, manually download https://dl.google.com/widevine-cdm/4.10.1440.19-win-x64.zip and replace extracted dll in ..\Kodi\cdm\ solved playback issue.

Similar problem is on *elec and as i read, similar solution works there (download fresh Acer Chromebook Tab 10 recovery image and manually extract and replace libwidevinecdm.so).

So for Windows, there should be in settings option to choose between current and other version (choose from list of versions.txt)
For arm, there should be option for re-download fresh version from recovery.

**Comment by mediaminister** [comment]
**Manual fix for Linux x64**
```
wget -qO- -O tmp.zip https://dl.google.com/widevine-cdm/4.10.1440.19-linux-x64.zip && unzip -o -j tmp.zip libwidevinecdm.so -d ~/.kodi/userdata/addon_data/script.module.inputstreamhelper/cdm && rm tmp.zip
```

**Comment by dagwieers** [comment]
I think we need to improve inputstreamhelper to make it more reliable/versatile:
- Get information on the current version being used (useful when reporting/troubleshooting)
- Make it possible to force a reinstallation (in case of troubleshooting)
- Select a specific version to be used (to test if a previous version still worked after upgrade)
- Keep previous versions (and switch back to a previous version without reinstallation)
- Log the various steps and the file(s) it downloads (so people can tinker away themselves)

**Comment by emilsvennesson** [comment]
Definitely. There's a lot of room for improvement in this regard. I haven't been able to work on this for a while though, and I'm not sure if I will be able to for the foreseeable future.

I'll happily review pull requests though if anyone feels like they're up for the task! 😉

**Comment by cache-sk** [comment]
I have quickfix too :)

With bool in settings (to use versions) and action in settings to force reinstall (inspired by pullrequest 52)

Tested on win64, with small flaw - after bool change you must first store settings..

https://github.com/cache-sk/script.module.inputstreamhelper

**Comment by dagwieers** [comment]
@cache-sk I would like to have a proper settings page and an API.

The settings would have the following categories and options:
**Actions**
- Install latest Widevine CDM (ask to reinstall if already installed)
- Manually select Widevine CDM version (lists online and on-disk versions)
- Remove Widevine CDM
**InputStream Adaptive**
- Install InputStream Adaptive
  - InputStream Adaptive settings...
**About**
*Shows information on installed Widevine CDM version or installed ChromeOS recovery image version*

The API would have the options:
- plugin://script.module.inputstreamhelper/widevine/install/latest
- plugin://script.module.inputstreamhelper/widevine/install/manual
- plugin://script.module.inputstreamhelper/widevine/remove

**Comment by dagwieers** [comment]
**Manual command line fix for ARM**
```
sed -i -e "s/'SPRING'/'SKATE'/g" ~/.kodi/addons/script.module.inputstreamhelper/lib/config.py && rm -f ~/.kodi/userdata/addon_data/script.module.inputstreamhelper/cdm/widevine_config.json
```

**Wrong libwidevinecdm.so**
```
kodi03:~ # md5sum ~/.kodi/cdm/libwidevinecdm.so
541c86d612adf2fe085ea409abe6c50e  /storage/.kodi/cdm/libwidevinecdm.so
```

**Correct libwidevinecdm.so**
```
kodi03:~ # md5sum ~/.kodi/cdm/libwidevinecdm.so
41c94b9ffa735fe4f412b7e9283dd2ff  /storage/.kodi/cdm/libwidevinecdm.so
```

**Comment by matthuisman** [comment]
http://k.mjh.nz/.decryptmodules/modules.v2.json

**Comment by matthuisman** [comment]
For anyone in a jam:
http://k.mjh.nz/.decryptmodules/modules.v2.json
http://k.mjh.nz/.decryptmodules/widevine/

**Comment by mediaminister** [comment]
I created a pull request to automatically update widevine when inputstreamhelper is updated and running for the first time https://github.com/emilsvennesson/script.module.inputstreamhelper/pull/61

**Comment by matthuisman** [comment]
> I created a pull request to automatically update widevine when inputstreamhelper is updated and running for the first time #61

The way I do it in mine is I create a string that contains
**system + arch + kodi_version + inputstream addon version**

Once installed, I store that string in the inputstream addon settings.

Every time it's run, it checks if those two are different, if yes - then reinstall.

If they are the same, i then check if it's more than 24 hours since last check.
If so, then do a silent reinstall.

As I store the md5 of the drm files online, it's a simple json file to download to check for new version. It's behind cloudflare cache so fast no matter where it is. 

Its probably quick enough to check everytime widevine is invoked to be honest.

**Comment by dagwieers** [comment]
@emilsvennesson Can we get #60 and #61 released ASAP given the scope and impact?

**Comment by mediaminister** [comment]
> I'll happily review pull requests though if anyone feels like they're up for the task! wink

@emilsvennesson How will this work for new releases? Should we open a pull request at https://github.com/xbmc/repo-scripts ourselves and ask you for approval or will you find time to handle this process yourself?

**Comment by basrieter** [comment]
Usually we only accept updates in the Kodi Repository from the original developer. In rare cases another Dev submits the PR but then the original author needs to approve.

**Comment by basrieter** [comment]
@emilsvennesson to get this done as fast as possible the two PR's (#60 and #61) should be merge into this repo and it's version should be bumped here. After that, somebody else could do the PR for the Kodi repo if that is OK?

**Comment by emilsvennesson** [comment]
I'm currently abroad so I can't cut a new release right now. You can PR a new release. I'll happily merge it and let you submit it to the Kodi repo.

**Comment by dagwieers** [comment]
Given that inputstreamhelper is a critical piece in the Kodi offering these days (for many important add-ons), I would like to discuss this project's governance at some point in the future. I count at least 4 active contributors which is a good foundation for a close collaboration. And the momentum is here now to make this happen.

**Comment by mediaminister** [comment]
@emilsvennesson I created the PR with the 0.3.5 release https://github.com/emilsvennesson/script.module.inputstreamhelper/pull/67

**Comment by matthuisman** [comment]
Here is the thread of where this add-on was born from:
**https://forum.kodi.tv/showthread.php?tid=321165**

I suggested way back then that Aussie Addon developers should have been the ones to create the official add-on as they already had a basis and where the most knowledgeable on the widevine stuff.

You can see there walking emilsvennesson through the best way to go about it and pointing him to the existing examples.

It was my view then, and still is now - that someone who knew what they were doing (Aussie addon guys etc) should be the ones who create the add-on.

IMO emilsvennesson just copied their code, and was the first to do a pull-request to kodi repo.
Then didn't give them any credit.

> I think @glennguy 's module would be pretty much ready to go into the official repo if he wanted to.

And this is what happens.
Other add-ons rely on an official add-on that gets pretty much abandoned.

**Comment by emilsvennesson** [comment]
@matthuisman I believe we've had this discussion before. Other add-ons relies/relied on third party sites redistributing the Widevine binary which breaches the EULA.

> IMO emilsvennesson just copied their code, and was the first to do a pull-request to kodi repo.
Then didn't give them any credit.

Never did I claim to have come up with the idea. I did it it in a way that would comply with the EULA and Kodi rules which the others do not. It's funny that you're claiming I "copied code" because they don't even remotely work in the same way.

I just wanted to get the ball rolling when I started this. @mediaminister @dagwieers You both have contributed a lot lately and is probably more familiar with the code than I am at this point. I'll happily make you collaborators if you feel like you're up for it! 😀

**Comment by mediaminister** [comment]
Thanks, please add me as collaborator. Let's move forward!
I'm confident @dagwieers also wants to improve this project as a collaborator

**Comment by basrieter** [comment]
I would also be willing to collaborate. And being a Kodi team member, it would make sense to help out with such an essential add-on. Together with @dagwieers and @mediaminister it would then be a dev team of better size!

**Comment by emilsvennesson** [comment]
I've sent you all invites! 🙂

**Comment by dagwieers** [comment]
@emilsvennesson Can you enable the Wiki-space so I can transfer some of the ideas there, and we can use Wiki-pages to collaborate on design ideas?

**Comment by nicoolaro** [comment]
can someone write manual commands I should type on RPI (running DietPi) for this to work?
I'm complete amateur on linux systems

**Comment by dagwieers** [comment]
You need to update your **inputstreamhelper** addon to **version 0.3.5**.
This should happen automatically within 24 hours of its release.

If you want to manually verify/update this package, first make sure you have the latest repository updates:
- First go to **Add-ons** » **My add-ons** » **Add-on repository** » **Kodi Add-on repository**
- Right-click on it (context-menu), select **Check for update** and click on it

With some luck your Kodi will now update all packages that have updates, to verify:
- Go to the **Information** page of an add-on that requires **inputstreamhelper** (right-click add-on, then select **Information**).
- Then go to **Dependencies**, and select **InputStream Helper**, click on it
- Verify that version **0.3.5** is already installed
  - If it is not installed, go to the **Update** button and click on it

**Comment by nicoolaro** [comment]
@dagwieers thanks so much, i follow the steps inputstreamhelper updated to 0.3.3 not 0.3.5, anyway i started HBOGO, asked me to download Chrome Os Recovery image, i did.
Everything works even though I'm on 0.3.3

Awesome!

**Comment by ErikBrown2** [comment]
@dagwieers Are your provided steps for the Pi supposed to fix the widevine issue? I have InputStream Helper version 0.3.5 but still my Pi with LibreElec and Retrospect does not play the NPO start programs. Is there anything else that I can do?

**Comment by basrieter** [comment]
You need to install the libraries again using the option from the Retrospect add-on settings. In Retrospect 5.0.1 (https://github.com/xbmc/repo-plugins/pull/2484) it will be done automagically.

**Comment by ErikBrown2** [comment]
@basrieter What do you mean with installing the libraries again? I have tried to manually update all dependencies of Retrospect as described by dagwieers but that did not solve the issue. Then I uninstalled Retrospect and reinstalled it. That also did not solve it. Is there another thing I can do short of reinstalling LibreElec?

**Comment by dagwieers** [comment]
@ErikBrown2 Normally installing Retrospect v0.3.5 will automatically trigger a new Widevine CDM download on the next playback. However, I have noticed (only when troubleshooting) that a new Widevine CDM only worked after a reboot. So could you try to reboot?

If this appears a general issue, we may want to look into why this is happening.
Do we have to remove some of the persistent information when updating the Widevine CDM live?

Also, if we can detect that the used Widevine CDM is revoked, we probably should check for an update as well. In that case there is no need to update the Widevine CDM on upgrade. I'll open a feature request for this.

**Comment by basrieter** [comment]
@dagwieers That is not correct. I don't use the inputstreamhelper at all in previous Retrospect versions. the user needs to manually do that via the Retrospect `add-on settings -> General -> Install Widevine using Inputstream helper`

**Comment by dagwieers** [comment]
@basrieter Ah ok, this is specific to Retrospect, I was imagining the more general case. Maybe this discussion ought to go in the Retrospect forums then.

**Comment by peno64** [comment]
I did the reinstall of Widevine using Inputream helper and it works on my PI. Before the update the stream just didn't play. After the update, the system hang but after a reboot it does work perfect. I tried a program from vrt nu. Note that that same program played without any widevine update on the vrt.nu addon.

Besides that and I know it is out of topic here. Will Widevine also install on the new RPI4?

Thx

**Comment by cache-sk** [comment]
FYI: there is fatal flaw with autoupdate only on first run. If user cancels it (i don't have time for this now), he stays with older widevine. There should be user action to redownload at all..

**Comment by dagwieers** [comment]
@basrieter Could this have been the issue you have seen (not auto-updating) ?

**Comment by dagwieers** [comment]
@cache-sk Maybe best to open a new issue for this.

**Comment by mediaminister** [comment]
@cache-sk I know autoupdate on first run can and will fail in some circumstances but it was the easiest and fastest way to fix this issue.
The purpose of version 0.3.5 was to fix Widevine CDM as fast as possible for as many users as possible. I didn't take the time to fully investigate the best possible method to update Widevine for all these different Kodi platforms.

Please open a new issue for this.

**Comment by dagwieers** [comment]
@cache-sk Please try #91, it allows you to access the inputstreamhelper settings and remove/install Widevine CDM.

**Comment by dagwieers** [comment]
With #193 merged for v0.5.0, we have everything discussed in this thread implemented.
Thanks to everyone who helped out!
